### PR TITLE
updated to include a runner flag which allows remote selenium to be used

### PR DESCRIPTION
**Pull request by nwc1991**
We needed to extend scholar-runner so that we can run it over a selenium server within a Docker environment.

We added a runner option where you can specific where to run the tests e.g. local, remote or broswerstack, we left the broswerstack flag in just so that we don't create a braking change, and added the code to run with a remote selenium server

### CODE REVIEW

**Comment by alexnaish** [comment]
Wouldn't worry about breaking changes, this is a nice feature and it would go out as a new major release. I'd rather keep things neat and tidy rather than add in extra features too haphazardly.

I'm not going to accept this just yet, I think it should be refactored to reduce an overlap in options and functionality. Essentially all the browserstack option does is set up a remote selenium session with the browserstack selenium location pre-specified - with a few extra options admittedly but they could be moved into the config file.

For this to be merged this needs:
- Consolidation of the CLI options / functionality.
  - Remove the --browserstack option in favour of the runner parameter.
  - Update the runner flag to accept either 'local' or 'remote', with it defaulting to local.
- `connectToRemoteSeleniumAndRunTests` and `setupBrowserStackAndRunTests` are essentially the same function duplicated.
- Move remote selenium configuration out of the function into config options.
- Refactor the previously mentioned functions into one `setupRemoteSessionAndRunTests` (for example).

**Review by nowarkdev — CHANGES_REQUESTED** [review]
The options object should be cleaner, the user and key settings shouldn't be hardcoded to be the BROWSERSTACK process variables. These should just be set to the relevant config fields. Since the config variable is a JS file, end users can choose to use process variables where required. 

Additionally, on `lib/runner.js:66` options.browserstackLocal will never be set as the option has been removed so this needs to be fixed.

File: lib/runner.js [hunk 1]
```diff
@@ -63,12 +63,11 @@ function setupRemoteSessionAndRunTests(config, options) {
             browserName: options.browser,
             platform: "ANY"
         };
-    browserOptions['browserstack.local'] = options.browserstackLocal;
     let bsOptions = {
         "logLevel": options.verbose ? "verbose" : "silent",
         "desiredCapabilities": browserOptions,
-        "user": process.env.BROWSERSTACK_USER,
-        "key": process.env.BROWSERSTACK_KEY,
+        "user": config.browserstack ? config.browserstack.user : '',
```
Comment (nowarkdev) [inline lib/runner.js]: How about the below, as it isn't directly related to browserstack. Any remote selenium server can have a username and password associated with it.

"user": config.user,
"key": config.key

File: lib/runner.js [hunk 2]
```diff
@@ -63,12 +63,11 @@ function setupRemoteSessionAndRunTests(config, options) {
             browserName: options.browser,
             platform: "ANY"
         };
-    browserOptions['browserstack.local'] = options.browserstackLocal;
     let bsOptions = {
```
Comment (nowarkdev) [inline lib/runner.js]: Could this not be done with something along the lines of an `Object.assign(standardOptions, config)`?

File: README.md [hunk 3]
```diff
@@ -86,6 +86,8 @@ An Example browserstack config:
 
     {
         browserstack: {
+            user: 'browserstack-user-name',
+            key: 'browserstack-automation-key',
```
Comment (nowarkdev) [inline README.md]: See above comment around fields.

**Review by nowarkdev — CHANGES_REQUESTED** [review]

**Comment by alexnaish** [comment]
Looks good from the diff, will checkout and give it a run through.

**Pull request merged**